### ISSUE TITLE
Find an alternative solution for respecting GPU isolation on a worker

### ISSUE THREAD

**Issue by epicfaace** [codalab/codalab-worksheets#2661]
#2415 was an attempted fix to respect GPU isolation on a worker.

Unfortunately, it caused GPUs to stop being available. We attempted to fix this issue by adding nvidia-smi to the worker Docker image (#2650), but that didn't fix it, so we had to revert it in #2649.

We should find out a solution to the above issue. Additionally, if such a solution does not require nvidia-smi to be directly available on the worker Docker image, we should revert the changes in #2650 as well.

**Comment by bkgoksel** [comment]
I thought I had implemented this in 2019: https://github.com/codalab/codalab-worksheets/pull/1015
But there might a part of it that I'm missing, but AFAIK I could run Codalab workers that would only use the GPUs slurm had assigned to them.

**Comment by nelson-liu** [comment]
> I thought I had implemented this in 2019: #1015

Yeah, that would definitely work. I had an open PR for using this at https://github.com/codalab/codalab-worksheets/pull/2351 , but we decided to try to make it a feature of any worker (https://github.com/codalab/codalab-worksheets/pull/2351#issuecomment-644486583). I think trying to get #2351 in and fixing the slurmworkermanager would be a good start.

More broadly, I think the issue is one with `--gpuset` defaults. the default (`ALL`) will get _all_ the GPUs on the host, regardless if the worker has access to them or not. I think we want to change this such that it gets only the GPUs that the worker has access to.

**Comment by nelson-liu** [comment]
Should we leave this open? This is fixed for the isolated case of the slurmworkermanager, but it isn't fixed if you do:

```
nlprun -g 1
cl-worker
```

(the `nlp-cl-worker` script has some bash magic that sets the GPUSET from the nvidia-smi environment, but we should and could fold that in to the worker itself). I also don't think the worker respects `CUDA_VISIBLE_DEVICES`

**Comment by percyliang** [comment]
Ok, sure that makes sense.
